src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs

```diff
@@ -195,6 +196,29 @@ public virtual SecurityToken CreateToken(SecurityTokenDescriptor tokenDescriptor
             return new Saml2SecurityToken(assertion);
         }
 
+        /// <inheritdoc/>
+        public override Task<TokenValidationResult> ValidateTokenAsync(string token, TokenValidationParameters validationParameters)
+        {
+            try
+            {
+                var claimsPrincipal = ValidateToken(token, validationParameters, out var validatedToken);
+                return Task.FromResult(new TokenValidationResult
+                {
+                    SecurityToken = validatedToken,
+                    ClaimsIdentity = claimsPrincipal?.Identity as ClaimsIdentity,
+                    IsValid = true,
+                });
+            }
+            catch (Exception ex)
+            {
+                return Task.FromResult(new TokenValidationResult
+                {
+                    IsValid = false,
+                    Exception = ex
+                });
+            }
```

Comment:
This is outside the scope of this PR but I'd like to see us as a team get away from this pattern.

I may be misunderstanding something but I believe we moved to this return-value-houses-the-exception-no-throw model because there was a request to increase performance. We're using exceptions here as glorified goto statements which is not great IMO.

The only exceptions, as I understand, that we'd actually want to expose here are of type SecurityTokenException so it seems like we could at least only catch that. (Very unclear to me what we expect our callers to do if they get an exception of an unrelated-to-Wilson type). 

The above is the first problem, the second problem is that we've just moved the perf cost of exception handling down the stack into our library instead of in the partner libraries which I'm not certain is much of an improvement. We could instead decide to pass the relevant exceptions back through the stack, this would almost certainly be more performant as well as avoiding the awkward poor-man's goto nature of the current pattern.

Reply:
Yeah this makes sense. We should only catch the validation exceptions, like SecurityTokenValidationException, and set them in the returned result.

Reply:
@TimHannMSFT if we assume that any application using our SDK is catching exceptions (as they should), we should have a discussion about changing our model.

Reply:
Thanks Danny I had another look at this after our discussion. This approach still seems good to me. I think in the future, we can even take a look at refactoring how JsonWebTokenHandler::ValidateSignature bubbles up exceptions. We throw from that level only to catch here which I think we can eventually approve but let's save that for later and not delay more here, nothing prevents us from addressing that later.

I would recommend adding a remarks explaining the exception behavior of the new method though, explaining it won't ever throw SecurityTokenExceptions and the caller is responsible for checking the response but that it can throw input validation exceptions etc.

Reply:
Added comments on the exceptions that can be thrown and the properties a caller should check to verify the validation result.